Error handling docs (#7575)
sveltejs/kit: documentation/docs/30-advanced/25-errors.md

---
title: Errors
---

Errors are an inevitable fact of software development. SvelteKit handles errors differently depending on where they occur, what kind of errors they are, and the nature of the incoming request.

### Error objects

> SvelteKit doesn't expose errors to users directly, Instead, errors are represented as simple `{ message: string }` objects by default.

Reviewer: I'm guessing the errors talked about here are from `throw new Error` statements, so maybe add "thrown" to clarify (vs let's say a 500 which is an error but not `Error`). Also, there's a comma here that I think should be a period
```suggestion
SvelteKit doesn't expose thrown errors to users directly. Instead, errors are represented as simple `{ message: string }` objects by default.
```

Reviewer 2: Trying to square the circle:

```suggestion
SvelteKit distinguishes between two kinds of errors - expected and unexpected ones. They are explained in more detail in the following sections. Both are represented as simple `{ message: string }` objects by default.
```

Author: Yeah I like that. Slightly more compact version — I think the 'Expected errors' heading is close enough that we can omit the 'explained below' bit:

```suggestion
SvelteKit distinguishes between expected and unexpected errors, both of which are represented as simple `{ message: string }` objects by default.
```

	if (!post) {
		throw error(404, {
			message: 'Not found'
		});
	}

	return { post };
}
```

This tells SvelteKit to set the response status code to 404 and render an [`+error.svelte`](/docs/routing#error) component, where `$page.error` is the object provided as the second argument to `error(...)`.

```svelte
/// file: src/routes/+error.svelte
<script>
	import { page } from '$app/stores';
</script>

<h1>{$page.error.message}</h1>
```

You can add extra properties to the error object if needed...

```diff
throw error(404, {
	message: 'Not found',
+	code: 'NOT_FOUND'
});
```

...otherwise, for convenience, you can pass a string as the second argument:

```diff
-throw error(404, { message: 'Not found' });
+throw error(404, 'Not found');
```

### Unexpected errors

An _unexpected_ error is any other exception that occurs while handling a request. Since these can contain sensitive information, unexpected error messages and stack traces are not exposed to users.

By default, unexpected errors are printed to the console (or, in production, your server logs), while the error that is exposed to the user has a generic shape:

```json
{ "message": "Internal Error" }
```

> You can add your own error handling — for example, sending errors to a reporting service, or returning a custom error object — using the [`handleError`](/docs/hooks#shared-hooks-handleerror) hook.

Reviewer 2: ```suggestion
Unexpected errors will go through the [`handleError`](/docs/hooks#shared-hooks-handleerror) hook, where you can add your own error handling — for example, sending errors to a reporting service, or returning a custom error object. 
```

Author: You can add code blocks inside suggestions, you just need to use four backticks on the outer fence (````` ````suggestion `````)

````suggestion
Unexpected errors will go through the [`handleError`](/docs/hooks#shared-hooks-handleerror) hook, where you can add your own error handling — for example, sending errors to a reporting service, or returning a custom error object. 

```js
/// file: src/hooks.server.js
// @errors: 2322 2571
// @filename: ambient.d.ts
const Sentry: any;

// @filename: index.js
// ---cut---
/** @type {import('@sveltejs/kit').HandleServerError} */
export function handleError({ error, event }) {
// example integration with https://sentry.io/
Sentry.captureException(error, { event });

return {
message: 'Whoops!',
code: error.code ?? 'UNKNOWN'
};
}
```
````

### Responses

If an error occurs inside `handle` or inside a [`+server.js`](/docs/routing#server) request handler, SvelteKit will respond with either a fallback error page or a JSON representation of the error object, depending on the request's `Accept` headers.

You can customise the fallback error page by adding a `src/error.html` file:

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<meta charset="utf-8" />
		<title>%sveltekit.error.message%</title>
	</head>
	<body>
		<h1>My custom error page</h1>
		<p>Status: %sveltekit.status%</p>
		<p>Message: %sveltekit.error.message%</p>
	</body>
</html>
```

> 

Reviewer 2: ```suggestion

SvelteKit will replace `%sveltekit.status%` and `%sveltekit.error.message%` with their corresponding values.

```

```diff
throw error(404, {
	message: 'Not found',
+	code: 'NOT_FOUND'
});
```

...otherwise, for convenience, you can pass a string as the second argument:

```diff
-throw error(404, { message: 'Not found' });
+throw error(404, 'Not found');
```

### Unexpected errors

An _unexpected_ error is any other exception that occurs while handling a request. Since these can contain sensitive information, unexpected error messages and stack traces are not exposed to users.

By default, unexpected errors are printed to the console (or, in production, your server logs), while the error that is exposed to the user has a generic shape:

```json
{ "message": "Internal Error" }
```

Unexpected errors will go through the [`handleError`](/docs/hooks#shared-hooks-handleerror) hook, where you can add your own error handling — for example, sending errors to a reporting service, or returning a custom error object. 

```js
/// file: src/hooks.server.js
// @errors: 2322 2571
// @filename: ambient.d.ts
const Sentry: any;

// @filename: index.js
// ---cut---
/** @type {import('@sveltejs/kit').HandleServerError} */
export function handleError({ error, event }) {

> // example integration with https://sentry.io/
> Sentry.captureException(error, { event });
> 
> return {
> message: 'Whoops!',
> code: error.code ?? 'UNKNOWN'
> };

Author: whoops i messed up the indentation somehow

```suggestion
	// example integration with https://sentry.io/
	Sentry.captureException(error, { event });

	return {
		message: 'Whoops!',
		code: error.code ?? 'UNKNOWN'
	};
```
